pkg/region/cache.go

```diff
@@ -0,0 +1,178 @@
+
+	once sync.Once
+}
+
+// expireKeyIfRequired expires a key if it's ready to be expired
+func (c *cacheStore) expireKeyIfRequired(cloud CloudName, r Name, v *cacheEntry) {
+	if time.Now().UTC().After(v.ExpiresAt) {
+		c.cloudsMu.Lock()
+		defer c.cloudsMu.Unlock()
+
+		delete(c.Clouds[cloud], r)
+	}
+}
+
+// get returns a cache entry for a given cloud/region pairing
+func (c *cacheStore) get(cloud CloudName, r Name) (*cacheEntry, bool) {
+	c.cloudsMu.RLock()
+	defer c.cloudsMu.RUnlock()
+
+	if _, ok := c.Clouds[cloud]; ok {
+		return nil, false
+	}
+
+	v, ok := c.Clouds[cloud][r]
+	if !ok {
+		return nil, ok
+	}
+
+	return &v, ok
+}
+
+// Get returns the duration of a cloud/region pairing, if it exists.
+// Otherwise ok is returned as false and time.Duration is it's zero-value.
+func (c *cacheStore) Get(cloud CloudName, r Name) (time.Duration, bool) {
+	c.once.Do(c.load) // read the cache from disk at most once
+
+	v, ok := c.get(cloud, r)
+	if !ok {
+		// not ok, so just return it's not there
+		return time.Duration(0), false
+	}
+
+	c.expireKeyIfRequired(cloud, r, v) // expire the key if required to do so
+
+	return v.Duration, ok
+}
+
+func (c *cacheStore) set(cloud CloudName, r Name, dur time.Duration) {
+	c.cloudsMu.Lock()
+	defer c.cloudsMu.Unlock()
+
+	if _, ok := c.Clouds[cloud]; !ok {
+		c.Clouds[cloud] = make(map[Name]cacheEntry)
+	}
+
+	c.Clouds[cloud][r] = cacheEntry{
+		Duration: dur,
+		// Expire in 8 hours
+		ExpiresAt: time.Now().UTC().Add(time.Hour * 8),
+	}
+}
+
+// Set sets the duration of a cloud/region pairing
+func (c *cacheStore) Set(cloud CloudName, r Name, dur time.Duration) error {
+	c.once.Do(c.load) // read the cache from disk at most once
+
+	c.set(cloud, r, dur) // set the key into our datastore
+
+	// save the result to disk
+	return c.save()
+}
+
+func (c *cacheStore) getCacheFilePath() (string, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	storageDir := RegionCachePath
+	if !filepath.IsAbs(storageDir) {
+		storageDir = filepath.Join(homeDir, storageDir)
+	}
+
+	return filepath.Join(storageDir, RegionCacheFile), nil
+}
+
+// load retrieves the cache from disk, if it exists, otherwise
+// it is returned uninitialized
+func (c *cacheStore) load() {
+	c.fileMu.Lock()
+	defer c.fileMu.Unlock()
+
+	// ensure that the cache is empty, or initialized if called
+	// on first run
+	c.Clouds = make(map[CloudName]map[Name]cacheEntry)
+
+	cacheFilePath, err := c.getCacheFilePath()
+	if err != nil {
+		return
+	}
+
+	f, err := os.Open(cacheFilePath)
+	if err != nil {
+		return
+	}
+
+	// lock the underlying datastore structure while we're decoding
+	// the file into it
+	c.cloudsMu.Lock()
+	defer c.cloudsMu.Unlock()
+
+	_ = json.NewDecoder(f).Decode(&c) //nolint:errcheck // Why: function signature/acceptable
+}
+
+// save saves the cache to disk
+func (c *cacheStore) save() error {
+	c.fileMu.Lock()
+	defer c.fileMu.Unlock()
+
+	cacheFilePath, err := c.getCacheFilePath()
```

Comment:
why not use the mutex in `getCacheFilePath`? i was always taught to only hold a lock for as long as you need it, and no longer. you're holding this lock for the duration of the function when you really only need it for line 163. you're not doing any more reading after that your locks on 174-175 are correct though

Reply:
I'm more used to putting locks/unlocks at the top of a body, but you're right it's not needed in there so I'm going to just move it below

Reply:
I would favor easy to maintain, correct code over performant code -- atleast until there is a demonstrable reason for perf.  In this case, I'd just use a single mutex and lock at the top of all exported functions -- this is quite easy to see that there is no possibility of deadlocks.  I'd also stick to a single lock per struct for that reason -- very easy to see that the code works correctly.   With multiple locks, code reviews can easily miss correctness problems and concurrency tests are very hard to get right.

Reply:
Yeah, that's why I tend to do that method, which ensures I don't accidentally mess up locks. That's why I made a `set` and `get` in the follow-up commits earlier. In this case, because I'm calling `defer` right after I'm less concerned about accidentally deadlocking, so I'll keep @george-e-shaw-iv's method for now. Still agree on usually keeping at the top, though.

Reply:
For example, we now have two mutexes and a sync.Once in one struct -- I think it is very hard for me to prove that this code is correct.   Even worse, if someoone makes changes, it is easy to miss the interactions between these as the diff will only have the snippet that changed.

I can prove the code is correct from concurrency p.o.v if there was exactly one mutex, all exported functions had the lock/defer-unlock block at the top, unexported functions were only called by exported functions and they had no other locks.

Reply:
If you must user multiple locks, I suggest creating embedded structs that scope each lock down:

```golang
type MainStruct struct {
      scope1 struct {
            mu sync.Mutex
            fields protected by mu
      }
      scope2 struct {
            mu sync.Mutex
            fields protected by mu
      }
}
```

This can be validated more easily by confirming that all access to fields within a scope are only acquired with mutexes on that scope.

Reply:
Updated to use one mutex again and be more clear. I still prefer the `c.once.Do(c.load)` for loading though, that's reading better to me because it's in the "friendly" `Get` and `Set` interfaces only. That work for you?

Reply:
(reasoning behind going back to one was because they are really tied together for the most part anyways)